torch/ao/quantization/fx/_model_report/detector.py

```diff
@@ -577,17 +600,214 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
+
+        for fqn, module in model.named_modules():
+            # check to see if module is of a supported type
+            is_supported_type: bool = (
+                sum(list(map(lambda x: type(module) is x, self.SUPPORTED_MODULES))) > 0
+            )
+
+            # if module is supported and it has a pre-observer
+            if is_supported_type and hasattr(module, self.DEFAULT_PRE_OBSERVER_NAME):
+                # we don't need actual observer, just the module weights
+                # calculate min and max vals
+                min_val, max_val = torch.aminmax(module.weight, dim=self.ch_axis)
+
+                # flatten entries since conv can have multiple dimensions
+                min_val = torch.flatten(min_val)
+                max_val = torch.flatten(max_val)
+
+                weight_info[fqn] = {
+                    self.PER_CHANNEL_MAX_KEY: max_val,
+                    self.PER_CHANNEL_MIN_KEY: min_val,
+                    self.GLOBAL_MAX_KEY: max(max_val),
+                    self.GLOBAL_MIN_KEY: min(min_val),
+                }
+
+        return weight_info
+
+    def _calculate_range_ratio(self, info_dict: Dict) -> torch.Tensor:
+        r"""
+        Takes in an info dict and calculates the s_c matrix.
+
+        Args:
+            info_dict (dict): A dictionary of either input or weight range info
+
+        Returns a tensor of values, where each value is the s_c stat for a different channel
+        """
+        # calculate the ratios of the info
+        per_channel_range = info_dict[self.PER_CHANNEL_MAX_KEY] - info_dict[self.PER_CHANNEL_MIN_KEY]
+        global_range = info_dict[self.GLOBAL_MAX_KEY] - info_dict[self.GLOBAL_MIN_KEY]
+
+        # if global range is 0, throw error
+        if global_range == 0:
+            raise ValueError("The range of the info dict is 0")
+
+        ratio = per_channel_range / global_range
+
+        return ratio
+
+    def _generate_comparision_values(self, input_info: Dict, weight_info: Dict) -> Dict[str, torch.Tensor]:
+        r"""
+        Takes in the information on the min and max values of the inputs and weights and:
+            Calculates the comp stat for each channel: s_c = sqrt(w_c/W)/sqrt(i_c/I)
+
+        Args:
+            input_info (dict): A dict mapping each observer to input range information
+            weight_info (dict): A dict mapping each observer to weight range information
+
+        Returns a dict mapping relavent observer fqns (str) to a 1-D tensor.
+            Each value is a different s_c value for a different channel
+        """
+        # create return dictionary for each observer
+        module_fqn_to_channel: Dict[str, torch.Tensor] = {}
+
+        # for each module (both passed in dicts should have same keys)
+        for module_fqn in input_info:
+
+            # raise error if not in weight info
+            if module_fqn not in weight_info:
+                raise KeyError("Both input_info and weight_info should have same keys")
```

Comment:
again, could probably use better message to the user, i.e. could not find weight for module X rather than a technical error message.